File: LeetCode/Medium/Even Odd Tree/Go/even_odd_tree.go
package main

import "fmt"

type TreeNode struct {
	Val int
	Left,
	Right *TreeNode
}

func New(val int) *TreeNode { return &TreeNode{Val: val, Left: nil, Right: nil} }

func even_odd_tree(root *TreeNode) bool {
	i := 0

	type Pair struct {
		node *TreeNode
		lvl  int
	}

	q := []Pair{{root, i}}
	for len(q) > 0 {
		n := len(q)
		if i%2 == 0 {
			for j := 0; j < n; j++ {
				curr := q[j]
				if (curr.node.Val%2 != 1) || (j > 0 && q[j].node.Val <= q[j-1].node.Val) {
					return false
				}
				if curr.node.Left != nil {
					q = append(q, Pair{node: curr.node.Left, lvl: curr.lvl + 1})
				}
				if curr.node.Right != nil {
					q = append(q, Pair{node: curr.node.Right, lvl: curr.lvl + 1})
				}
			}
		} else {
			for j := 0; j < n; j++ {
				curr := q[j]
				if (curr.node.Val%2 != 0) || (j > 0 && q[j].node.Val >= q[j-1].node.Val) {
					return false
				}
				if curr.node.Left != nil {
					q = append(q, Pair{node: curr.node.Left, lvl: curr.lvl + 1})
				}
				if curr.node.Right != nil {
					q = append(q, Pair{node: curr.node.Right, lvl: curr.lvl + 1})
				}
			}
		}
		i++
		q = q[n:]
	}
	return true
}

func main() {
	root := New(5)
	root.Left = New(4)
	root.Right = New(2)
	root.Right.Left = New(7)
	root.Left.Left = New(3)
	root.Left.Right = New(3)
	// root.Left.Left.Left = New(12)
	// root.Left.Left.Right = New(8)
	// root.Right.Right = New(9)
	// root.Right.Left = New(7)
	// root.Right.Right.Right = New(2)
	// root.Right.Left.Left = New(6)
	fmt.Println(even_odd_tree(root))
}

File: LeetCode/Medium/Even Odd Tree/c++/even_odd_tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isEvenOddTree(TreeNode* root) {
        bool level = true; //true even level 
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            int n = q.size();
            queue<TreeNode*> temp;
            for (int i = 0; i < n-1; i++) {
                TreeNode *curr = q.front();
                q.pop();
                TreeNode *next = q.front();
                if (level) {
                    if (curr->val >= next->val || curr->val % 2 == 0 || next->val % 2 == 0) return false;
                } else {
                    if (curr->val <= next->val || curr->val % 2 != 0 || next->val % 2 != 0) return false;
                }
                if (curr->left != nullptr) temp.push(curr->left);
                if (curr->right != nullptr) temp.push(curr->right);
            }
            if (level && !q.empty() && q.front()->val % 2 == 0) return false;
            if (!level && !q.empty() && q.front()->val % 2 != 0) return false;
            if (!q.empty() && q.front()->left != nullptr) temp.push(q.front()->left);
            if (!q.empty() && q.front()->right != nullptr) temp.push(q.front()->right);
            level = !level;
            q = temp;
        }
        return true;
    }
};
